Fix: prioritize VE.Direct network power

schlimmchen opened the pull request:
the DPL shall use the VE.Direct network input power to derive (guesstimate) the available solar output. we did this before already, but a regression was introduced with the solar charger refactoring.

we now also try to be smarted about guesstimating the output: we reduce the network total input power by the input power of all charge controllers we are listening to and account their battery output power instead. the remainder of the network input power, i.e., the input power of all networked charge controllers that we are not listening to, is corrected for the average efficiency and added to the return value.

Status: Not tested yet. Can't test tomorrow, this needs to wait until Sunday.

Closes #1793.

AndreasBoehm commented on src/solarcharger/victron/Stats.cpp:
```diff
@@ -33,6 +34,41 @@ std::optional<float> Stats::getOutputPowerWatts() const
+    // the efficiency of the networked controllers that we are not wired to.
+    std::vector<float> efficiencies;
+    std::optional<float> networkPower = std::nullopt;
+    int16_t accountedInputPower = 0;
+    int16_t accountedOutputPower = 0;
```
I doubt that we will reach 32 kW power with any of the setups

AndreasBoehm replied:
But i think that we can use `uint16_t` instead of `int16_t`

schlimmchen replied:
Not really required. Assuming these values will not grow larger than 32kW is reasonable.

However, since we are doing float calculations anyways, we might as well use floats here.

schlimmchen replied:
I did not see your comments, Andreas, until I reloaded the page after I was notified by email...

> But i think that we can use uint16_t instead of int16_t

No, that won't work (for `accountedOutputPower`) as the battery output power can be negative for devices with a load output.

AndreasBoehm commented on src/solarcharger/victron/Stats.cpp:
```diff
@@ -33,6 +34,41 @@ std::optional<float> Stats::getOutputPowerWatts() const
+        *networkPower = std::max<float>(0, *networkPower);
+
+        // average efficiency of all controllers we are wired to
+        float efficiency = std::accumulate(efficiencies.begin(), efficiencies.end(), 0.0f) / efficiencies.size();
+
+        return accountedOutputPower + (*networkPower * efficiency / 100.0f);
```
this is already covered by the check `if (networkPower.has_value()) {`

schlimmchen replied:
Yeah, well... Good idea, but not really applicable: The whole program path where the efficiency variable is calculated from the vector is not executed if we don't have a value for `networkPower`. If that variable has a value, we also have at least one value in the vector, see the loop above.

Leaves the situation where the vector has value(s) but they are zero. That can actually happen! As far as I can tell right now from looking at the code it *will* happen during the night.

![image](https://github.com/user-attachments/assets/4c1f09d1-9d5a-46a7-8e33-8aede913b5a5)

Indeed. Good catch, copilot :blush:

schlimmchen replied:
Wait... I misunderstood. The efficiency value itself being zero is actually not an issue. If that is the case, e.g., during the night, the product of efficiency and and the unaccounted network input power will be zero, which is fine.

copilot's issue with a possible division by zero is only relevant if the vector is empty, as we are using `vector::size()` as a divisor. That's always at least one if we have a value for `oNetworkPower`, and since the use of `vector::size()` is guarded by having a value for `oNetworkPower`, we are good.

AndreasBoehm commented:
I put this on my productive system this morning, so far it looks good.

schlimmchen commented:
My [test protocol](https://github.com/user-attachments/files/19614804/2025-04-05_test_protocol.pdf) (change of plans, need to do the test today rather than tomorrow) shows plausible values for all scenarios that I tested, so I am confident in this change.

schlimmchen commented:
Renamed `networkPower` to `oNetworkPower`, changed type of `accounted*Power` variables to float, added comment, moved declaration of sum near the last loop.

SW-Niko commented:
Hallo @schlimmchen,
ich denke es gibt noch ein Problem das auftritt wenn jemand den Load-Ausgang verwendet. 🤔 
Dein Test-Protokoll habe ich mir angeschaut, aber dieser theoretische Fall mit negativer Leistung fehlt.

![grafik](https://github.com/user-attachments/assets/d6dfd5aa-9a3d-4ec9-882b-8ba77777de7e)

Wenn ich mit den Daten aus dem Bild in die Funktion Stats::getOutputPowerWatts() gehe dann liefert sie 120W zurück. Das ist jetzt theoretisch berechnet und die Verluste habe ich weggelassen.
Aber 70W wäre das richtige Ergebnis.

Der Fehler entsteht an der Stelle, wo negative Teilmengen von POut vor der Addierung auf 0 gesetzt werden.
Mathematisch korrekt müssten wir erst die Summen bilden und ganz zum Schluss prüfen, ob der Rückgabewert negativ ist.

Aber …. ich würde negative Werte an dieser Stelle nicht ausblenden. Der Rückgabewert von Stats::getOutputPowerWatts() wird auch im LiveView angezeigt und ich denke wenn der negativ ist dann soll er hier auch negativ Angezeigt werden und nicht 0. Die Überprüfung auf < 0 sollte erst erfolgen, wenn der Wert für weitere Berechnungen verwendet wird.

![grafik](https://github.com/user-attachments/assets/ba4782a8-df0f-4173-ad14-c5c0072e1f83)

Die MPPT benötigen beim Start etwas mehr Energie, dann springt der Wert bei mir von 0W auf -2W

Des weiteren sollten wir noch darauf hinweisen, das der Load Output vom einem MPPT nur dann richtig erfasst wird wenn zu dem MPPT eine VE.Direct Verbindung besteht. Ansonsten ist das Ergebnis falsch und wir können nichts dagegen machen.

Wie ist deine Meinung dazu?

SW-Niko commented:
Ja das passt. 
Aber ich denke Schlimmchen hat das gemacht, weil woanders im Code ein negativer Rückgabewert zu einem Problem führt.
Ich schau mir das nochmal genauer an wo das Problem war.

AndreasBoehm commented:
Ich hab einen PR erstellt: https://github.com/hoylabs/OpenDTU-OnBattery/pull/1850

schlimmchen commented:
@SW-Niko @AndreasBoehm Danke :heart:
